[FIX] make sure the thanks button of the newsletter snippet is hidden before save

Guillaume-gdi opened the pull request:
Before this commit the "Thanks" button of the newsletter block could be
saved as visible. This meant that a visitor who did not subscribe to the
newsletter would see the "Thanks" button on loading before it
disappeared. This commit solves this problem by hiding the button before
saving.

opw-2802139
--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/91539).

qsm-odoo commented on addons/web_editor/static/src/js/wysiwyg/widgets/link_dialog.js:
```diff
@@ -48,6 +48,7 @@ var LinkDialog = Dialog.extend({
             {type: this.isButton ? 'link' : '', label: _t("Link"), btnPreview: 'link'},
             {type: 'primary', label: _t("Primary"), btnPreview: 'primary'},
             {type: 'secondary', label: _t("Secondary"), btnPreview: 'secondary'},
+            {type: 'success', label: _t("Success"), btnPreview: 'success'},
```
@Guillaume-gdi We don't want to suggest "success" as a new button type, that's actually something we have been wanting to discourage in the past and try to make users only use "primary", "secondary" or "link" buttons (even though we added the "custom" button possibility in 15.0, we still show "primary" etc as first suggestions).

qsm-odoo commented:
After last review, nothing was worth considering fixing in 14.0 for this success button edition problem. A new PR was created for 15.0 fix instead of continuing on this one: https://github.com/odoo/odoo/pull/91630

This current PR will however be kept to handle another problem found during review: allowing to save the "Thanks" button of the newsletter snippet as visible on page load.

Guillaume-gdi commented:
@qsm-odoo I'm having trouble finding a good fix for this... The "Thanks" button is saved even if the newsletter block options have not started. So the only solution I see is to add the d-none when saving the element(actual PR) which doesn't seem to be the right fix... I wonder if there is a better way to do it..?

qsm-odoo commented:
@Guillaume-gdi Like all other widgets, it should restore its DOM properly when entering edit mode. IMO, entering edit mode should thus hide the "Thanks" button, empty the input, etc. It does not make sense to me that if you are subscribed to your own newsletter, you cannot edit the subscribe button anymore.

Then, if you want to edit the "Thanks" button, I guess the editor should provide a button to switch between the "Subscribe" and the  "Thanks" state (like the Countdown widget which has a button to switch to the state where the countdown is finished so that the users can edit the shown message). Easy to do in 14.0, maybe less in 13.0. If you don't find a way, retarget this to 14.0 and ignore that for 13.0. It's not critical and there are workarounds (even if annoying ones).

Guillaume-gdi commented:
@qsm-odoo This is ready; I also added a commit because I realized that when we modified the newsletter block and the reCaptcha was not installed we were spammed with notifications.

qsm-odoo commented on addons/website_mass_mailing/static/src/js/website_mass_mailing.js:
```diff
@@ -88,6 +88,9 @@ publicWidget.registry.subscribe = publicWidget.Widget.extend({
             return def;
         }
 
+        if (this.editableMode) {
+            return this._super.apply(this, arguments);
+        }
```
You are now calling _super twice...

qsm-odoo commented on addons/website_mass_mailing/static/src/js/website_mass_mailing.editor.js:
```diff
@@ -64,6 +71,40 @@ options.registry.mailing_list_subscribe = options.Class.extend({
             self.getParent()._onRemoveClick($.Event( "click" ));
         });
     },
+    /**
+     * @override
+     */
+    cleanForSave() {
+        if (!this.$target[0].matches('.o_newsletter_popup')) {
+            this.$target[0].querySelector('.js_subscribe_btn').classList.remove('d-none');
+            this.$target[0].querySelector('.js_subscribed_btn').classList.add('d-none');
+        }
+    },
```
Why duplicating the code of the public widgets, can't you let it do it?

qsm-odoo commented on addons/website_mass_mailing/static/src/js/website_mass_mailing.editor.js:
```diff
@@ -54,6 +54,13 @@ options.registry.mailing_list_subscribe = options.Class.extend({
+    /**
+     * @see this.selectClass for parameters
+     */
+    toggleShowThanks(previewMode, widgetValue, params) {
+        this.$target[0].querySelector('.js_subscribe_btn').classList.toggle('d-none');
+        this.$target[0].querySelector('.js_subscribed_btn').classList.toggle('d-none');
+    },
```
Using the same class as the public widget (d-none) to toggle the visibility of those buttons, don't you have the same problem that was solved for the countdown snippet with https://github.com/odoo/odoo/commit/c37354d457f5b868673b4f974e401f4c635062d2 as mentioned earlier ?

qsm-odoo commented on addons/website_mass_mailing/static/src/js/website_mass_mailing.editor.js:
```diff
@@ -64,6 +71,40 @@ options.registry.mailing_list_subscribe = options.Class.extend({
+    _computeWidgetState: function (methodName, params) {
+        if (methodName === 'toggleShowThanks') {
+            const subscribeBtnEl = this.$target[0].querySelector('.js_subscribe_btn');
+            return subscribeBtnEl && subscribeBtnEl.classList.contains('d-none');
```
I don't like the implicit cast, _computeWidgetState is expected to return a string, return a string

qsm-odoo commented on addons/website_mass_mailing/static/src/js/website_mass_mailing.editor.js:
```diff
@@ -64,6 +71,40 @@ options.registry.mailing_list_subscribe = options.Class.extend({
+    _renderCustomXML(uiFragment) {
+        const btnEl = document.createElement('we-checkbox');
+        btnEl.setAttribute('string', _t("Display Thanks Button"));
+        btnEl.dataset.toggleShowThanks = true;
+        btnEl.dataset.noPreview = true;
```
Same, dataset values are string, write `'true'`, not `true`

Guillaume-gdi commented:
@qsm-odoo This one is ready for your review

qsm-odoo commented on addons/website_mass_mailing/i18n/website_mass_mailing.pot:
```diff
@@ -108,6 +108,13 @@ msgstr ""
+#. module: website_mass_mailing
+#. openerp-web
+#: code:addons/website_mass_mailing/static/src/js/website_mass_mailing.editor.js:0
+#, python-format
```
To update .pot file, you should:
- Install the module with your branch
- Go to settings -> translation -> export translation
- Export the pot file by choosing "empty translation" (something like that) + PO type
- Replace the content of the git .pot file
- Only update what's related to your branch (there are probably many mistakes made by other but better not to noise our fixes PR)

Here I think you made a mistake doing it by hand, but I may be wrong?

Guillaume-gdi replied:
Yes BSO had given me this useful tip few weeks ago, I double checked and it's ok :+1:

qsm-odoo replied:
Ok, I thought the `#, python-format` stuff was not added in this case, my bad then :grimacing:

qsm-odoo commented on addons/website_mass_mailing/static/src/js/website_mass_mailing.editor.js:
```diff
@@ -64,6 +76,42 @@ options.registry.mailing_list_subscribe = options.Class.extend({
+     * @override
+     */
+    cleanForSave() {
+        if (!this.$target[0].matches('.o_newsletter_popup')) {
```
There is a dedicated class to handle the newsletter popup... so don't mention the newsletter popup class in the base class :grimacing:  Here I think you can just remove that `if`?

qsm-odoo commented on addons/website_mass_mailing/static/src/js/website_mass_mailing.js:
```diff
@@ -91,24 +90,51 @@ publicWidget.registry.subscribe = publicWidget.Widget.extend({
+    /**
+     * Modifies the elements to have the view of a subscriber/non-subscriber.
+     *
+     * @param  {Object} data
```
I see a double space after `@param`

qsm-odoo commented on addons/website_mass_mailing/static/src/js/website_mass_mailing.js:
```diff
@@ -91,24 +90,51 @@ publicWidget.registry.subscribe = publicWidget.Widget.extend({
+        const thanksBtnEl = this.$target[0].querySelector('.js_subscribed_btn');
+        const emailInputEl = this.$target[0].querySelector('input.js_subscribe_email');
+
+        subscribeBtnEl.toggleAttribute('disabled', isSubscriber);
```
Do you actually need `toggleAttribute`? Can't you just do `.disabled = ` ? It would just change the property and not the DOM markup, just like before.

qsm-odoo commented on addons/website_mass_mailing/static/src/scss/website_mass_mailing_popup.scss:
```diff
@@ -20,3 +20,14 @@
         opacity: 1;
     }
 }
+// TODO: Put the following rules in a file that is not sent to the visitor.
+body.editor_enable {
+    .s_newsletter_subscribe_form {
+        .o_enable_preview {
+            display: block !important;
+        }
+        .o_disable_preview {
+            display: none !important;
+        }
+    }
+}
```
Detail, but since you will need to modify this PR a bit: as a stable fix, I would prefer you put these rules in a website editor file (even though it's a website_mass_mailing rule). That way you don't impact stable version visitors. Totally a detail in this case, but just so you know.

Note: in 15.0, you would be able to add a new file, so there won't ever be that kind of problem anymore anyway.

qsm-odoo commented on addons/website_mass_mailing/static/src/js/website_mass_mailing.editor.js:
```diff
@@ -155,6 +155,11 @@ options.registry.newsletter_popup = options.registry.mailing_list_subscribe.exte
      * @override
      */
     start: function () {
```
Commit message: "again on the newsletter popup again to see them again." -> 3x "again" in ~10words ^^ But this commit does not makes sense on its own: you introduce the "Display Thanks Button" option in the previous commit and add a third one to fix it :thinking: Just squash?

qsm-odoo commented on addons/website_mass_mailing/static/src/js/website_mass_mailing.editor.js:
```diff
@@ -155,6 +155,11 @@ options.registry.newsletter_popup = options.registry.mailing_list_subscribe.exte
      * @override
      */
     start: function () {
+        this.$target.on('shown.bs.modal.newsletter_popup_option', () => {
+            this.trigger_up('activate_snippet', {
+                $snippet: this.$target,
+            });
+        });
```
I don't understand this, that would need comments?  Is not it already activated once the modal is opened?

Guillaume-gdi commented:
@qsm-odoo I applied the changes and squashed the last two commits

robodoo commented:
Merge method set to rebase and fast-forward

qsm-odoo commented:
@robodoo r+
